### PR TITLE
cpu manager: specify the container CPU set during the creation

### PR DESCRIPTION
**Pull request by cynepco3hahue**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
We can set the container `cpuset.cpus` during the creation and it
will not need to call to update resources after the container creation.

The additional side effect of the change, that the runc process that is responsible
to create the container will run with the same CPU affinity because the
runc runs on the cpuset provided in the config.json arg.

It will prevent undesirable interrupts on isolated CPUs.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

Signed-off-by: Artyom Lukianov <alukiano@redhat.com>

### CODE REVIEW

**Comment by klueska** [comment]
I haven't looked closely at this yet, but can you elaborate on what bug this is fixing (i.e. you marked it as `/kind bug`) and what exactly you mean by:

> The additional side effect of the change, that the runc process that is responsible to create the container will run with the same CPU affinity because the runc runs on the cpuset provided in the config.json arg

Does the new hook you implement make sure that the list of allocated CPUs get passed to the runc config?
Does this work for all versions of runc?

**Comment by cynepco3hahue** [comment]
> I haven't looked closely at this yet, but can you elaborate on what bug this is fixing (i.e. you marked it as `/kind bug`) and what exactly you mean by:
> 
> > The additional side effect of the change, that the runc process that is responsible to create the container will run with the same CPU affinity because the runc runs on the cpuset provided in the config.json arg

Hi Kevin, I wasn't totally sure about the kind, maybe setting it as cleanup will be more correct. But one thing for sure, once we are running some latency-sensitive workload under our environment it possible that some other container that crashed at the same time cause runc to be scheduled on the isolated CPU(caught via `ftrace`). We talked with one of the runc maintainers and he confirmed that the runc creation process runs on top of the CPU set that you passed to the container config during the creation. Currently, under the kubelet, we have the next flow creation, `cpumanager.Allocate -> create container -> update container resources(update cpuset.cpus) -> start container`, so in the current flow, the runc creation process runs on top of all available CPUs under the node because update the CPU set too late.

> 
> Does the new hook you implement make sure that the list of allocated CPUs get passed to the runc config?
> Does this work for all versions of runc?

Yes, it passes cpu set during the container creation so the CRI already gets the updated config. Regarding the second question, IIUC each Kubernetes version goes with the specific runc version, so in general, it should work only for runc that go together with the latest Kubernetes. It is hard to verify because the runc creation process exists a very short amount of time under the node.

File: pkg/kubelet/cm/internal_container_lifecycle.go [hunk 1]
```diff
@@ -37,13 +38,23 @@ type internalContainerLifecycleImpl struct {
...
 
+func (i *internalContainerLifecycleImpl) PreCreateContainer(pod *v1.Pod, container *v1.Container, containerConfig *runtimeapi.ContainerConfig) error {
+	if containerConfig.Linux != nil && containerConfig.Linux.Resources != nil {
+		if i.cpuManager != nil {
+			allocatedCPUs := i.cpuManager.GetAllocatedResources(pod, container)
+			if allocatedCPUs != "" {
+				containerConfig.Linux.Resources.CpusetMems = allocatedCPUs
+			}
+		}
+	}
```
Comment (klueska) [inline pkg/kubelet/cm/internal_container_lifecycle.go:49]: I don't think we simply want to check for `containerConfig.Linux != nil && containerConfig.Linux.Resources != nil` and then set CPUs in them if they exist. I think we want to explicitly create those structures if they do not exist and then add CPUs to them. Moreover, this should explicitly only be done on linux, so something like:
```
if i.cpuManager == nil {
    return nil
}

if runtime.GOOS != "linux" {
    return nil
}

allocatedCPUs := i.cpuManager.GetAllocatedResources(pod, container)
if allocatedCPUs == "" {
    return nil
}

if containerConfig.Linux == nil {
    containerConfig.Linux = ...
}

if containerConfig.Linux.Resources == nil {
    containerConfig.Linux.Resources = ...
}

containerConfig.Linux.Resources.CpusetMems = allocatedCPUs

return nil
```

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 2]
```diff
@@ -63,7 +63,10 @@ type Manager interface {
 	// AddContainer is called between container create and container start
 	// so that initial CPU affinity settings can be written through to the
 	// container runtime before the first process begins to execute.
-	AddContainer(p *v1.Pod, c *v1.Container, containerID string) error
+	AddContainer(p *v1.Pod, c *v1.Container, containerID string)
+
+	// GetAllocatedResources return the CPU affinity for the container
+	GetAllocatedResources(pod *v1.Pod, container *v1.Container) string
```
[the remark this answers is not in the document]
Comment (cynepco3hahue) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go:69]: I just wanted to have an interface method with the neutral name because I will want to have the same for the memory manager, WDT?

File: pkg/kubelet/cm/internal_container_lifecycle.go [hunk 3]
```diff
@@ -37,13 +38,23 @@ type internalContainerLifecycleImpl struct {
 	topologyManager topologymanager.Manager
 }
 
+func (i *internalContainerLifecycleImpl) PreCreateContainer(pod *v1.Pod, container *v1.Container, containerConfig *runtimeapi.ContainerConfig) error {
```
[the remark this answers is not in the document]
Comment (cynepco3hahue) [reply pkg/kubelet/cm/internal_container_lifecycle.go:41]: Good catch! done

File: pkg/kubelet/cm/internal_container_lifecycle.go (hunk 1 above)
Comment (cynepco3hahue) [reply pkg/kubelet/cm/internal_container_lifecycle.go:49]: I checked again the code and I can see that all needed structure initialized under https://github.com/kubernetes/kubernetes/blob/3b38f6a08e255b3eb2e34ffb3c3cecacc081f334/pkg/kubelet/kuberuntime/kuberuntime_container.go#L160 so I will just remove all redundant checks.

File: pkg/kubelet/cm/cpumanager/cpu_manager.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go:69]: I still think we should just consolidate it down to a single call of:
```
GetCPUs(podUID, containerName string) cpuset.CPUSet
```

We should then expand `cpuset.CPUSet` to include helper functions for:
```
ToSliceInt64() []int64
ToSliceNoSortInt64() []int64
```

Then the two call sites that use it can each call functions directly attached to the `CPUSet` type to get the format that they want.

This should be added as a standalone commit before the rest of the changes here.

[the remark this answers is not in the document]
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go:69]: I'm happy to rename the function to `GetAllocatedCPUs()` if you like, but I think it's better to keep it CPU specific.

File: pkg/kubelet/cm/internal_container_lifecycle.go [hunk 4]
```diff
@@ -37,13 +38,22 @@ type internalContainerLifecycleImpl struct {
...
 	if i.cpuManager != nil {
-		err := i.cpuManager.AddContainer(pod, container, containerID)
-		if err != nil {
-			return err
+		allocatedCPUs := i.cpuManager.GetAllocatedResources(pod, container)
+		if allocatedCPUs != "" {
+			containerConfig.Linux.Resources.CpusetMems = allocatedCPUs
 		}
 	}
```
Comment (klueska) [inline pkg/kubelet/cm/internal_container_lifecycle.go:47]: Does something else protect this from getting called on non.linux systems? Is creation of the CPUManager somehow only limited to linux?

Comment (klueska) [reply pkg/kubelet/cm/internal_container_lifecycle.go:47]: In fact, I see this in `pkg/kubelet/cm/container_manager_windows.go`
```
func (cm *containerManagerImpl) InternalContainerLifecycle() InternalContainerLifecycle {
    return &internalContainerLifecycleImpl{cpumanager.NewFakeManager(), topologymanager.NewFakeManager()}
}
```
Which will obviously cause problems on windows since I'm guessing it doesn't support `containerConfig.Linux.Resources.CpusetMems`.

Comment (cynepco3hahue) [reply pkg/kubelet/cm/internal_container_lifecycle.go:47]: We have two implementations of the container manager that compiled according to build flags, so if it windows specific build it will not include at all https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/internal_container_lifecycle.go

Comment (klueska) [reply pkg/kubelet/cm/internal_container_lifecycle.go:47]: I don't quite follow. The code I posted above clearly shows that the windows code instantiates an instance of `internalContainerLifecycleImpl`, passing it an instance of `cpumanager.NewFakeManager()`, meaning that `i.cpuManager != nil` will not be `nil` inside its call to `InternalContainerLifecycle()`.

File: pkg/kubelet/cm/cpumanager/cpu_manager.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cynepco3hahue) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go:69]: Do you want a separate PR for it or the separate commit will be good enough for it?

[the remark this answers is not in the document]
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go:69]: Separate commit is fine.

File: pkg/kubelet/cm/internal_container_lifecycle.go (hunk 4 above)
Comment (cynepco3hahue) [reply pkg/kubelet/cm/internal_container_lifecycle.go:47]: Oh, you are right, but in this case, FakeManager will return an empty string during the call to `GetAllocatedResources`, and the `if` check will not pass.

Comment (klueska) [reply pkg/kubelet/cm/internal_container_lifecycle.go:47]: Yeah. So it looks like it wouldn't cause issues in its current form, but it's also very subtle why it doesn't.
I would still either:
1. Add a `runtime.GOOS == "linux"` check here; or
2. Split the implementation of this one function into separate `_linux.go`, `_windows.go` files.

Comment (cynepco3hahue) [reply pkg/kubelet/cm/internal_container_lifecycle.go:47]: I can see that we have the same thing for the `AddContainer` can we leave such refactoring to the separate PR or do you prefer to finish with it under the current one?

Comment (klueska) [reply pkg/kubelet/cm/internal_container_lifecycle.go:47]: `AddContainer()` is OK because it calls back into the `CPUManager `to do its thing. The real CPUManager will add the container to its internal containerMap and the fake CPUManager will do nothin (neither of which is linux specific). The only reason we need to branch for linux here is because we explicitly set a  linux specific field in `containerConfig.Linux.Resources.CpusetMems`.  If/when windows supports something similar, they can implement this differently using the same underlying CPUmanager.

Comment (cynepco3hahue) [reply pkg/kubelet/cm/internal_container_lifecycle.go:47]: Makes sense, will create different files for Linux and windows implementation.

**Comment by cynepco3hahue** [comment]
@klueska I tested the PR locally and the reason why it fails under the CI, it because of the docker use. The fix works just fine with the container + runc.

**Comment by cynepco3hahue** [comment]
@klueska CPU manager test passed can you please review again?

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 5]
```diff
@@ -80,7 +80,7 @@ type Manager interface {
 
 	// GetCPUs implements the podresources.CPUsProvider interface to provide allocated
 	// cpus for the container
-	GetCPUs(podUID, containerName string) []int64
+	GetCPUs(podUID, containerName string) *cpuset.CPUSet
```
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go:83]: This should not be a pointer to a CPUSet, but just a CPUSet.
Under the hood it is just a struct wrapping a map, so in terms of storage, it is already just passing a pointer around.

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 6]
```diff
@@ -63,7 +63,7 @@ type Manager interface {
 	// AddContainer is called between container create and container start
 	// so that initial CPU affinity settings can be written through to the
 	// container runtime before the first process begins to execute.
-	AddContainer(p *v1.Pod, c *v1.Container, containerID string) error
+	AddContainer(p *v1.Pod, c *v1.Container, containerID string)
```
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go:66]: The comment here is now incorrect.

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 7]
```diff
@@ -481,11 +462,7 @@ func (m *manager) updateContainerCPUSet(containerID string, cpus cpuset.CPUSet)
...
-func (m *manager) GetCPUs(podUID, containerName string) []int64 {
-	cpus := m.state.GetCPUSetOrDefault(string(podUID), containerName)
-	result := []int64{}
-	for _, cpu := range cpus.ToSliceNoSort() {
-		result = append(result, int64(cpu))
-	}
-	return result
+func (m *manager) GetCPUs(podUID, containerName string) *cpuset.CPUSet {
+	cpus := m.state.GetCPUSetOrDefault(podUID, containerName)
+	return &cpus
```
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go:467]: Should not return a pointer type.

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 8]
```diff
@@ -481,11 +462,7 @@ func (m *manager) updateContainerCPUSet(containerID string, cpus cpuset.CPUSet)
 		})
 }
 
-func (m *manager) GetCPUs(podUID, containerName string) []int64 {
-	cpus := m.state.GetCPUSetOrDefault(string(podUID), containerName)
-	result := []int64{}
-	for _, cpu := range cpus.ToSliceNoSort() {
```
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go:487]: I see the old implementation of this used `ToSliceNoSort()`, but your updated use on https://github.com/kubernetes/kubernetes/pull/98019/files#diff-c07d343a8d459721656cf7ddacda71506bde01a5cac288d15a4aba6bfc9a1caeR1031 does not use a `NoSort()` variant.

File: pkg/kubelet/cm/cpumanager/fake_cpu_manager.go [hunk 9]
```diff
@@ -69,7 +69,7 @@ func (m *fakeManager) State() state.Reader {
 	return m.state
 }
 
-func (m *fakeManager) GetCPUs(podUID, containerName string) []int64 {
+func (m *fakeManager) GetCPUs(podUID, containerName string) *cpuset.CPUSet {
```
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/fake_cpu_manager.go:72]: Should not be pointer

File: pkg/kubelet/cm/cpumanager/state/state_checkpoint_test.go [hunk 10]
```diff
@@ -227,8 +227,8 @@ func TestCheckpointStateRestore(t *testing.T) {
 			restoredState, err := NewCheckpointState(testingDir, testingCheckpoint, tc.policyName, tc.initialContainers)
 			if err != nil {
 				if strings.TrimSpace(tc.expectedError) != "" {
-					if strings.Contains(err.Error(), "could not restore state from checkpoint") &&
-						strings.Contains(err.Error(), tc.expectedError) {
+					tc.expectedError = "could not restore state from checkpoint: " + tc.expectedError
+					if strings.HasPrefix(err.Error(), tc.expectedError) {
```
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/state/state_checkpoint_test.go:231]: This change seems unrelated and a bit sstrange to be updating `tc.expectedError` like this. What are you trying to achieve?

File: pkg/kubelet/cm/cpuset/cpuset.go [hunk 11]
```diff
@@ -198,6 +199,29 @@ func (s CPUSet) ToSliceNoSort() []int {
...
+}
+
+// ToSliceNoSortInt64 returns a slice of int64 that contains all elements from
+// this set.
+func (s CPUSet) ToSliceNoSortInt64() []int64 {
+	var result []int64
+	for cpu := range s.elems {
+		result = append(result, int64(cpu))
+	}
+
```
Comment (klueska) [inline pkg/kubelet/cm/cpuset/cpuset.go:221]: nit: unnecessary newline

File: pkg/kubelet/cm/cpuset/cpuset.go [hunk 12]
```diff
@@ -198,6 +199,29 @@ func (s CPUSet) ToSliceNoSort() []int {
...
 }
 
+// ToSliceInt64 returns an ordered slice of int64 that contains all elements from
+// this set
+func (s CPUSet) ToSliceInt64() []int64 {
+	var result []int64
+	for cpu := range s.elems {
+		result = append(result, int64(cpu))
+	}
+
```
Comment (klueska) [inline pkg/kubelet/cm/cpuset/cpuset.go:209]: nit: unnecessary newline

File: pkg/kubelet/cm/cpumanager/cpu_manager.go (hunk 5 above)
Comment (cynepco3hahue) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go:83]: I think the only overhead will be because of struct copy, why do you not like to use a pointer here?

File: pkg/kubelet/cm/cpumanager/cpu_manager.go (hunk 6 above)
Comment (cynepco3hahue) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go:66]: Goot catch, done

File: pkg/kubelet/cm/cpumanager/cpu_manager.go (hunk 7 above)
Comment (cynepco3hahue) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go:467]: done

File: pkg/kubelet/cm/cpumanager/cpu_manager.go (hunk 8 above)
Comment (cynepco3hahue) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go:487]: nice catch, done

File: pkg/kubelet/cm/cpuset/cpuset.go (hunk 11 above)
Comment (cynepco3hahue) [reply pkg/kubelet/cm/cpuset/cpuset.go:221]: done

File: pkg/kubelet/cm/cpuset/cpuset.go (hunk 12 above)
Comment (cynepco3hahue) [reply pkg/kubelet/cm/cpuset/cpuset.go:209]: done

File: pkg/kubelet/cm/cpumanager/state/state_checkpoint_test.go (hunk 10 above)
Comment (cynepco3hahue) [reply pkg/kubelet/cm/cpumanager/state/state_checkpoint_test.go]: done

File: pkg/kubelet/cm/cpumanager/cpu_manager.go (hunk 5 above)
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: Mostly just because the `CPUSet` type was designed so that you wouldn't have to pass around a pointer to it.

It's already passed around efficiently as the CPUSet type itself. Passing a pointer to it is actually *less*  efficient, because the exact same size object is being passed around, but there is one more level of indirection you have to go through get to the underlying map.

In other words, the definiton of the `CPUSet` struct looks as follows:
```
type CPUSet struct {
    elems map[int]struct{}
}
```

Which (in terms of storage) is essentially the same as the following, since go stores `map`s as pointers:
```
type CPUSet struct {
    elems uintptr
}
```

Which then (in terms of storage) is exactly the same as a standalone `uintptr`.

So by passing a full `CPUSet` type, you are essentially just passing a direct copy of the underlying `uintptr` to its embedded map (with 0 extra overhead). By passing a pointer to a `CPUSet` you are still passing a `uintptr` around, but you have to follow one more level of indirection before getting to the underlying map's pointer.

Comment (klueska) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: For reference:
https://play.golang.org/p/Eb0u32gTgDR

Comment (cynepco3hahue) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: I see thanks for the explanation.

Comment (klueska) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: For reference:
https://play.golang.org/p/WeCYE-5MY9O

Comment (klueska) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: For reference:
https://play.golang.org/p/TlclpQYkBLS

File: pkg/kubelet/kuberuntime/kuberuntime_container.go [hunk 13]
```diff
@@ -167,6 +167,13 @@ func (m *kubeGenericRuntimeManager) startContainer(podSandboxID string, podSandb
 		return s.Message(), ErrCreateContainerConfig
 	}
 
+	err = m.internalLifecycle.PreCreateContainer(pod, container, containerConfig)
+	if err != nil {
+		s, _ := grpcstatus.FromError(err)
+		m.recordContainerEvent(pod, container, "", v1.EventTypeWarning, events.FailedToCreateContainer, "Internal PreCreateContainer hook failed: %v", s.Message())
+		return s.Message(), ErrPreStartHook
```
Comment (klueska) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go:174]: Is `ErrPreStartHook` the correct error type to return here? Should we add a new `ErrPreCreateHook` error?

Comment (cynepco3hahue) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go]: True, done

**Comment by klueska** [comment]
Thanks for putting this together @cynepco3hahue.

With the introduction of https://github.com/kubernetes/kubernetes/pull/87759, the logic to allocate exclusive CPUs to a pod was moved from pod creation time to pod admission time. As such, the full set of exclusive CPUs that should be allocated to a pod is now known before the pod itself ever gets created. This now gives us the opportunity to set each container's `cpuset` *before* it gets launched.

In the old way of doing things this was not the case. Since we didn't actually allocate exclusive CPUs to a container until *after* it was created, it was necessary to make an `Update()` call against the runtime API in order to update the `cpuset` it was created with.

Now that we know the CPUSet before container creation this is no longer necessary, and we can update it before the container actually starts. This PR adds the logic to do this in a clean and consistent manner.

/lgtm
/approve

**Comment by cynepco3hahue** [comment]
@klueska A lot of thanks for the great review!

**Comment by klueska** [comment]
Unfortunately the small changes to:
```
pkg/kubelet/dockershim/helpers_linux.go
pkg/kubelet/kuberuntime/kuberuntime_container.go
```
mean that this is no longer isolated to just `pkg/kubelet/cm` (where I have approval rights).

We will need to get someone with approval rights at the `pkg/kubelet` level to review as well.

/cc @derekwaynecarr

**Comment by derekwaynecarr** [comment]
dockershim and kuberuntime changes all look good.

thanks for the thorough review @klueska 

/lgtm
/approve
